placement/algo/algo.go

```diff
@@ -0,0 +1,470 @@
+			tryHost := heap.Pop(ph.hostHeap).(*hostShards)
+			// do not place to same rack for now
+			if from != nil && tryHost.hostRack() == from.hostRack() {
+				sameRack = append(sameRack, tryHost)
+				continue
+			}
+			triedHosts = append(triedHosts, tryHost)
+			if ph.hostHeap.getTargetLoadForHost(tryHost.hostAddress())-len(tryHost.shardsSet) <= 0 {
+				// this is where "some" is, at this point the best host option in the cluster
+				// from a different rack has reached its target load, time to break out of the loop
+				break outer
+			}
+			if ph.canAssignHost(shard, from, tryHost) {
+				ph.assignShardToHost(shard, tryHost)
+				delete(shards, shard)
+				for _, triedHost := range triedHosts {
+					heap.Push(ph.hostHeap, triedHost)
+				}
+				break
+			}
+		}
+	}
+	for _, host := range sameRack {
+		heap.Push(ph.hostHeap, host)
+	}
+	for _, triedHost := range triedHosts {
+		heap.Push(ph.hostHeap, triedHost)
+	}
+}
+
+func (ph placementHelper) generatePlacement() placementSnapshot {
+	return placementSnapshot{uniqueShards: ph.uniqueShards, rf: ph.rf, shardsLen: ph.getShardLen(), hostShards: ph.hostShards}
+}
+
+// hostHeap provides an easy way to get best candidate host to assign/steal a shard
+type hostHeap struct {
+	hosts                 []*hostShards
+	rackToHostsMap        map[string]map[*hostShards]struct{}
+	targetLoad2           map[string]int
+	hostCapacityAscending bool
+}
+
+func newHostHeap(hosts []*hostShards, hostCapacityAscending bool, targetLoad map[string]int, rackToHostMap map[string]map[*hostShards]struct{}) *hostHeap {
+	hHeap := &hostHeap{hostCapacityAscending: hostCapacityAscending, hosts: hosts, targetLoad2: targetLoad, rackToHostsMap: rackToHostMap}
+	heap.Init(hHeap)
+	return hHeap
+}
+
+func (hh hostHeap) getTargetLoadForHost(hostAddress string) int {
+	return hh.targetLoad2[hostAddress]
+}
+func (hh hostHeap) Len() int {
+	return len(hh.hosts)
+}
+
+func (hh hostHeap) Less(i, j int) bool {
+	hostI := hh.hosts[i]
+	hostJ := hh.hosts[j]
+	leftLoadOnI := hh.getTargetLoadForHost(hostI.hostAddress()) - len(hostI.shardsSet)
+	leftLoadOnJ := hh.getTargetLoadForHost(hostJ.hostAddress()) - len(hostJ.shardsSet)
+	// if both host has tokens to be filled, prefer the one on a bigger rack
+	// since it tends to be more picky in accepting shards
+	if leftLoadOnI > 0 && leftLoadOnJ > 0 {
+		if hostI.hostRack() != hostJ.hostRack() {
+			return len(hh.rackToHostsMap[hostI.hostRack()]) > len(hh.rackToHostsMap[hostJ.hostRack()])
+		}
+	}
+	// compare left capacity on both hosts
+	if hh.hostCapacityAscending {
+		return leftLoadOnI > leftLoadOnJ
+	}
+	return leftLoadOnI < leftLoadOnJ
+}
+
+func (hh hostHeap) Swap(i, j int) {
+	hh.hosts[i], hh.hosts[j] = hh.hosts[j], hh.hosts[i]
+}
+
+func (hh *hostHeap) Push(h interface{}) {
+	host := h.(*hostShards)
+	hh.hosts = append(hh.hosts, host)
+}
+
+func (hh *hostHeap) Pop() interface{} {
+	n := len(hh.hosts)
+	host := hh.hosts[n-1]
+	hh.hosts = hh.hosts[0 : n-1]
+	return host
```

Comment:
yeah I was doing that because the heap examples seems to be doing that, but then the result needs to be nil-checked then cast, so I changed it to always do a hh.Len()>0 check before Pop(), not sure if it's still necessary to do nil check here